File: 0701-0800/0727-Minimum Window Subsequence/0727-Minimum Window Subsequence.py
class Solution:
    def minWindow(self, S: str, T: str) -> str:
        m, n = len(S), len(T)
        if m < n:
            return ''
        start = [-1] * (n + 1)
        start[0] = 0
        minLen, minStart = m + 1, -1
        for i in range(1, m + 1):
            for j in range(min(i, n), 0, -1):
                if S[i - 1] == T[j - 1]:
                    start[j] = start[j - 1]
            start[0] = i
            if start[n] != -1 and i - start[n] < minLen:
                minStart = start[n]
                minLen = i - start[n]
        
        return S[minStart: minStart + minLen] if minStart != -1 else ''

File: 0701-0800/0727-Minimum Window Subsequence/0727-Minimum Window Subsequence.cpp
class Solution {
public:
    string minWindow(string S, string T) {
        int m = S.size(), n = T.size();
        if (m < n) {
            return "";
        }
        vector<vector<int>> start(1 + m, vector<int>(1 + n, -1));
        int minLen = m + 1, minStart = -1;
        
        for (int i = 0; i <= m; ++i) {
            start[i][0] = i;
        }
        
        for (int i = 1; i <= m; ++i) {
            for (int j = 1; j <= min(i, n); ++j) {
                if (S[i - 1] == T[j - 1]) {
                    start[i][j] = start[i - 1][j - 1];
                }
                else {
                    start[i][j] = start[i - 1][j];
                }
            }
            
            if (start[i][n] != -1 && i - start[i][n] < minLen) {
                minStart = start[i][n];
                minLen = i - start[i][n];
            }
        }        
        
        return minStart != -1 ? S.substr(minStart, minLen) : "";
    }
};

// Optimize space
class Solution2 {
public:
    string minWindow(string S, string T) {
        int m = S.size(), n = T.size();
        if (m < n) {
            return "";
        }
        vector<int> start(1 + n, -1);
        int minLen = m + 1, minStart = -1;        
        start[0] = 0;
        for (int i = 1; i <= m; ++i) {
            for (int j = min(i, n); j >= 1; --j) {
                if (S[i - 1] == T[j - 1]) {
                    start[j] = start[j - 1];
                }
            }
            start[0] = i;
            if (start[n] != -1 && i - start[n] < minLen) {
                minStart = start[n];
                minLen = i - start[n];
            }
        }        
        
        return minStart != -1 ? S.substr(minStart, minLen) : "";
    }
};